// receipt_row/receipt_row.js
const STATE = require('STATE')
const statedb = STATE(__filename)
const { sdb, get } = statedb(fallback_module)

module.exports = receipt_row

async function receipt_row(opts = {}) {
  const { id, sdb } = await get(opts.sid)
  const { drive } = sdb

  const on = {
    style: inject,
    data: ondata,
    icons: iconject,
  }

  let dricons = []

  const el = document.createElement('div')
  const shadow = el.attachShadow({ mode: 'closed' })

  shadow.innerHTML = `
    <div class="receipt-row"></div>
    <style></style>
  `

  const style = shadow.querySelector('style')
  const row = shadow.querySelector('.receipt-row')

  await sdb.watch(onbatch)

  return el

  function fail(data, type) {
    throw new Error('invalid message', { cause: { data, type } })
  }

  async function onbatch(batch) {
    for (const { type, paths } of batch) {
      const data = await Promise.all(paths.map(path => drive.get(path).then(file => file.raw)))
      const func = on[type] || fail
      await func(data, type)
    }
  }

  function inject(data) {
    style.textContent = data[0]
  }

  function iconject(data) {
    dricons = data 
  }

  async function ondata(data) {
    let { label, value, link, icon } = data[0] || {}

    let value_html = value || ""

    if (link) {
      value_html = `<a href="${value}" target="_blank" class="receipt-link">${value}</a>`
      row.className = `receipt-row link`
    }
    
    if (icon) {
      if (icon == "btc.svg"){
          value_html = `<span class="receipt-icon">${dricons[1]}</span> ${value}`
          row.className = `receipt-row total`
      } else {
          value_html = `<span class="receipt-icon">${dricons[0]}</span> ${value}` // which would be our default icon it could be anything
      } 
    }
    row.innerHTML = `
      <div class="receipt-label">${label}</div>
      <div class="receipt-value">${value_html}</div>
      ${icon ? "" : `<div class="divider"></div>`}
    `
  }
}

function fallback_module() {
  return {
    api: fallback_instance
  }
  function fallback_instance(opts) {
    return {
      drive: {
        'icons/': {
          'default.svg': { '$ref': 'default.svg'},
          'btc.svg': { '$ref': 'btc.svg' },
        },
        'style/': {
         'style.css': {
            raw: `
              .receipt-row {
                padding: 8px 0;
                display: flex;
                flex-direction: column;
                gap: 5px;
              }

              .receipt-label {
                color: #6D6E6F;
                font-size: 14px;
                margin-bottom: 4px;
              }

              .receipt-value {
                color: #000;
                font-size: 18px;
              }

              .receipt-row.total  .receipt-value {
                font-size: 28px;
                font-weight: 600;
              }

              .receipt-row.link .receipt-value {
                color: #4479FF;
                text-decoration: none;
                cursor: pointer;
              }

              .divider {
                height: 1px;
                background: #ddd;
                margin-top: 6px;
              }

            `
          }
        },  
        'data/': {
          'opts.json': { raw: opts }
        }
      }
    }
  }
}
